contracts/nexus-gateway/src/contract/execute.rs

```diff
@@ -15,20 +14,46 @@ where
     S: Store,
 {
     pub fn route_messages(
-        mut self,
+        self,
         sender: Addr,
-        msgs: Vec<Message>,
+        msgs: Vec<msg::Message>,
     ) -> Result<Response<nexus::Message>> {
         match sender {
-            sender if sender == self.config.nexus => todo!(),
+            sender if sender == self.config.nexus => self
+                .route_to_router(
+                    msgs.into_iter()
+                        .map(TryInto::try_into)
+                        .collect::<Result<Vec<_>>>()?,
+                )
+                .change_context(ContractError::RouteToRouter),
             sender if sender == self.config.router => self
-                .route_to_nexus(msgs)
+                .route_to_nexus(msgs.into_iter().map(Into::into))
                 .change_context(ContractError::RouteToNexus),
             _ => Err(report!(ContractError::Unauthorized)),
         }
     }
 
-    fn route_to_nexus(&mut self, msgs: Vec<Message>) -> Result<Response<nexus::Message>> {
+    fn route_to_router(
+        self,
+        msgs: impl IntoIterator<Item = nexus::Message>,
```

Comment:
this is used in a very specific context. Consider changing the type to the simpler `Vec`?

Reply:
I also prefer `Vec<nexus::Message>` here, and doing the type conversion only once, at the call site.